Code/Javascript/objects-assignment-#6.js

```diff
@@ -0,0 +1,86 @@
+const restaurant = {
+    name: "Gigi's Pizza Shack",
+    address: "123 Main St, Portland OR 97200",
+    tags: ["pizza", "family", "dine-in", "take-out", "arcade"],
+    website: "http://www.gigispizza.com/",
+    staff: {
+      owner: {
+        name: "Gigi",
+        "phone number": "123-234-3456",
+      },
+      manager: {
+        name: "Rose",
+        "phone number": "234-345-4567",
+      },
+      chef: {
+        name: "Musa",
+        "phone number": "345-456-5678",
+      },
+    },
+  };
+
+  // Challenge #1: Accessing Data
+console.log("Restauraunt Name:", restaurant.name);
+console.log("Tag Name:", restaurant.tags[2]);
+console.log("The Chef's Name:", restaurant.staff.chef.name);
+
+restaurant.tags.forEach(tags =>{
+    console.log(tags);
+})
+
+  //Challenge #2: Updating Data
+
+  restaurant.address = "Victoria Plaza, Parklands Road, Nairobi Kenya"
+
+  restaurant.staff.hr ={
+    name: "Wonder",
+    "phone number": "07123456789",
+  }
+
+  delete restaurant.website
+
+  console.log(restaurant)
+
+
+  // Challenge #3 : Creating a new object
+  const menu = {
+    frenchfries: 200,
+    nyamachoma: 2000,
+    burger: 550,
+    ugalibeef: 300,
+
+   
+
+  //Challenge #4: Creating an object method
+
+  order(items){
+    let totalAmount = 0
+
+    items.forEach(items =>{
+        if(this.hasOwnProperty(items)) {
+            totalAmount += this[items]
+        }else{
+            console.log(`Item ${items} is not found in the menu.`);
+        }
+    })
+    return totalAmount
+}
+
+}
+
+console.log(menu.order(["ugalibeef", "burger","frenchfries"]));
+console.log(menu.order(["nyamachoma", "frenchfries"]));
+
+//Extra Challenges 
+
+restaurant.menu = {
+    frenchfries: 200,
+    nyamachoma: 2000,
+    burger: 550,
+    ugalibeef: 300,
+}
+
+console.log(restaurant)
+console.log(menu.order(["nyamachoma", "ugalibeef"]));
+
+
```

Comment:
You've created a comprehensive set of data and functionalities for the `restaurant` object and `menu` object. Let's review the code and make sure everything is functioning correctly, while also adding some improvements and explanations:

### Restaurant Object

This object holds information about the restaurant, including the name, address, tags, website, and staff details.

### Challenge #1: Accessing Data

```javascript
console.log("Restaurant Name:", restaurant.name);
console.log("Tag Name:", restaurant.tags[2]);
console.log("The Chef's Name:", restaurant.staff.chef.name);

restaurant.tags.forEach(tag => {
    console.log(tag);
});
```
- This section correctly accesses and prints various data from the `restaurant` object.
- Using `forEach` to iterate over tags and print each one is also correct.

### Challenge #2: Updating Data

```javascript
restaurant.address = "Victoria Plaza, Parklands Road, Nairobi, Kenya";

restaurant.staff.hr = {
    name: "Wonder",
    "phone number": "07123456789",
};

delete restaurant.website;

console.log(restaurant);
```
- The address is updated, and an HR staff member is added. The website property is deleted from the `restaurant` object.

### Challenge #3: Creating a New Object

The `menu` object is created to store the prices of items.

```javascript
const menu = {
    frenchfries: 200,
    nyamachoma: 2000,
    burger: 550,
    ugalibeef: 300,
```

### Challenge #4: Creating an Object Method

```javascript
order(items) {
    let totalAmount = 0;

    items.forEach(item => {
        if (this.hasOwnProperty(item)) {
            totalAmount += this[item];
        } else {
            console.log(`Item ${item} is not found on the menu.`);
        }
    });
    return totalAmount;
}
```
- The `order` method calculates the total cost of an order based on the items array.
- The method checks if each item exists in the `menu` object and sums the prices. If an item isn't found, it logs a message.

### Extra Challenges

Adding the `menu` object as a property of the `restaurant` object.

```javascript
restaurant.menu = {
    frenchfries: 200,
    nyamachoma: 2000,
    burger: 550,
    ugalibeef: 300,
};

console.log(restaurant);
console.log(menu.order(["nyamachoma", "ugalibeef"]));
```
- This section successfully adds the `menu` as a property of the `restaurant` object and prints it out.

### Improvements

1. **Avoiding Duplication**: If the `menu` is already defined separately, consider referencing it instead of creating a new object under `restaurant.menu`. This avoids duplication.

2. **Consistent Naming**: Use consistent and descriptive names for properties and methods.

3. **Edge Cases**: Consider handling edge cases such as empty item lists in the `order` method.

Here's a slightly improved version:

```javascript
const restaurant = {
    name: "Gigi's Pizza Shack",
    address: "123 Main St, Portland OR 97200",
    tags: ["pizza", "family", "dine-in", "take-out", "arcade"],
    staff: {
        owner: {
            name: "Gigi",
            "phone number": "123-234-3456",
        },
        manager: {
            name: "Rose",
            "phone number": "234-345-4567",
        },
        chef: {
            name: "Musa",
            "phone number": "345-456-5678",
        },
        hr: {
            name: "Wonder",
            "phone number": "07123456789",
        }
    },
    menu: {
        frenchfries: 200,
        nyamachoma: 2000,
        burger: 550,
        ugalibeef: 300,
    },
};

console.log("Restaurant Name:", restaurant.name);
console.log("Tag Name:", restaurant.tags[2]);
console.log("The Chef's Name:", restaurant.staff.chef.name);

restaurant.tags.forEach(tag => {
    console.log(tag);
});

restaurant.address = "Victoria Plaza, Parklands Road, Nairobi, Kenya";
delete restaurant.website;

console.log(restaurant);

const menu = {
    frenchfries: 200,
    nyamachoma: 2000,
    burger: 550,
    ugalibeef: 300,

    order(items) {
        let totalAmount = 0;

        items.forEach(item => {
            if (this.hasOwnProperty(item)) {
                totalAmount += this[item];
            } else {
                console.log(`Item ${item} is not found on the menu.`);
            }
        });
        return totalAmount;
    }
};

console.log(menu.order(["ugalibeef", "burger", "frenchfries"]));
console.log(menu.order(["nyamachoma", "frenchfries"]));

restaurant.menu = menu;

console.log(restaurant);
console.log(menu.order(["nyamachoma", "ugalibeef"]));
```

This version reduces redundancy by directly using the `menu` object and adds a more streamlined way of handling updates and logging.